File: examples/planet_positions.c
/* Example: compute planetary positions for today.
 * Compile: gcc -o planet_positions planet_positions.c -I./include -L./target/release -llibre_ephemeris -lm
 */
#include "le_ephemeris.h"
#include <stdio.h>
#include <math.h>

int main() {
    double jd_ut = le_julday(2026, 6, 22.5, 1);
    double xx[24];
    char serr[256];

    int flags = LE_FLG_XYZ | LE_FLG_J2000
        | LE_FLG_NOABERR | LE_FLG_NOGDEFL | LE_FLG_NOBIRR | LE_FLG_NONUT;

    printf("JD = %.1f (June 22, 2026)\n\n", jd_ut);

    int planets[] = {LE_MERCURY, LE_VENUS, LE_MARS, LE_JUPITER, LE_SATURN,
                     LE_URANUS, LE_NEPTUNE, LE_PLUTO};
    const char *names[] = {"Mercury", "Venus", "Mars", "Jupiter", "Saturn",
                          "Uranus", "Neptune", "Pluto"};

    for (int i = 0; i < 8; i++) {
        int rc = le_calc_ut(jd_ut, planets[i], flags, xx, serr);
        if (rc != LE_OK) {
            printf("%-8s: error %d\n", names[i], rc);
            continue;
        }
        double dist = sqrt(xx[LE_X]*xx[LE_X] + xx[LE_Y]*xx[LE_Y] + xx[LE_Z]*xx[LE_Z]);
        printf("%-8s: (%.4f, %.4f, %.4f)  dist=%.4f AU\n",
               names[i], xx[LE_X], xx[LE_Y], xx[LE_Z], dist);
    }

    // Moon position (geocentric)
    int rc = le_calc_ut(jd_ut, LE_MOON, flags, xx, serr);
    if (rc == LE_OK) {
        double dist = sqrt(xx[LE_X]*xx[LE_X] + xx[LE_Y]*xx[LE_Y] + xx[LE_Z]*xx[LE_Z]);
        printf("\nMoon (geocentric): (%.4f, %.4f, %.4f)  dist=%.4f AU = %.1f km\n",
               xx[LE_X], xx[LE_Y], xx[LE_Z], dist, dist * 149597870.7);
    }

    return 0;
}

File: examples/planet_positions.rs
/// Example: compute planetary positions at J2000.
/// Run with: cargo run --example planet_positions
use libre_ephemeris::calc::le_calc_ut;
use libre_ephemeris::constants;

fn main() {
    let jd_ut = 2451545.0; // J2000.0

    let flags = constants::LE_FLG_XYZ | constants::LE_FLG_J2000
        | constants::LE_FLG_NOABERR | constants::LE_FLG_NOGDEFL
        | constants::LE_FLG_NOBIRR | constants::LE_FLG_NONUT;

    let planets = [
        (constants::LE_MERCURY, "Mercury"),
        (constants::LE_VENUS, "Venus"),
        (constants::LE_MARS, "Mars"),
        (constants::LE_JUPITER, "Jupiter"),
        (constants::LE_SATURN, "Saturn"),
        (constants::LE_URANUS, "Uranus"),
        (constants::LE_NEPTUNE, "Neptune"),
    ];

    println!("Planetary positions at J2000.0 (barycentric J2000 equatorial):");
    for (ipl, name) in &planets {
        let mut xx = [0.0_f64; 24];
        let mut serr = [0_i8; 256];
        let rc = unsafe { le_calc_ut(jd_ut, *ipl, flags, xx.as_mut_ptr(), serr.as_mut_ptr()) };
        if rc == constants::LE_OK {
            let dist = (xx[0] * xx[0] + xx[1] * xx[1] + xx[2] * xx[2]).sqrt();
            println!("  {:<8}: ({:.6}, {:.6}, {:.6}) AU  dist={:.6} AU",
                     name, xx[0], xx[1], xx[2], dist);
        }
    }
}
